Release 5.23.0

dktapps opened the pull request:
This includes all changes from minor-next as of December 5th


## !!!! DO NOT squash merge this PR !!!!

ipad54 committed:
Implemented sound when drinking a potion (#6444)

Commit:
Merge 'stable' into 'minor-next'

Automatic merge performed by: https://github.com/pmmp/RestrictedActions/actions/runs/11769186885

Commit:
Merge 'stable' into 'minor-next'

Automatic merge performed by: https://github.com/pmmp/RestrictedActions/actions/runs/11802296296

dktapps committed:
Promise: allow zero promises
not supporting this has caused problems every time this function has been used in reality so far (#6092 and #6333).

Commit:
Merge 'stable' into 'minor-next'

Automatic merge performed by: https://github.com/pmmp/RestrictedActions/actions/runs/11824307499

IvanCraft623 committed:
Implement Goat horns (#5232)

Co-authored-by: ipad54 <63200545+ipad54@users.noreply.github.com>
Co-authored-by: Dylan T. <dktapps@pmmp.io>

dktapps committed:
Use reflection to locate BlockTypeIds and ItemTypeIds for VanillaBlocks/VanillaItems (#6498)

Use reflection to locate BlockTypeIds and ItemTypeIds for VanillaBlocks/VanillaItems

Since BlockTypeIds and ItemTypeIds are derived from VanillaBlocks and VanillaItems respectively anyway (they only exist to allow identifying blocks/items without having to create instances of them), this hack is probably OK, and reduces the chances of mistakes.
Previously it was explored to have these IDs generated by auto-incrementing in VanillaBlocks/Items and have the constants generated that way, but this proved to be too problematic because of unstable diffs no matter how we chose to sort the elements. See #6313 for previous research on the subject.

This is obviously not a desirable hack to keep long-term. In the future it will probably make sense to redesign VanillaBlocks like so:

enum VanillaBlocks { ... }
VanillaBlocks::STONE (the type ID)
VanillaBlocks::STONE->new() (to create a block)

However, more research is needed on this, as I'd prefer not to make block creation any more verbose.

dktapps committed:
Merge branch 'stable' of github.com:pmmp/PocketMine-MP into minor-next

dktapps committed:
Update Language dependency

dktapps committed:
Merge branch 'minor-next' of github.com:pmmp/PocketMine-MP into minor-next

ipad54 committed:
Extract glow lichen multi face logic into traits (#6396)

This will be useful for future block additions

ipad54 committed:
Implemented vanilla /xp command (#6429)

Commit:
Adjust Sugar Cane to break when there is no water (#6486)

Commit:
Merge 'stable' into 'minor-next'

Automatic merge performed by: https://github.com/pmmp/RestrictedActions/actions/runs/11862895039

dktapps committed:
start.sh: print warnings on unusual exit codes from the server process (#6497)

dktapps committed:
PlayerInteractEvent: added APIs to independently control reaction of item and block (#4683)

This allows, for example, banning the usage of spawn eggs, without preventing opening of doors, without the need for item ID whitelists.

It also allows customizing the behaviour of item and block interactions when sneaking - it's now possible to force spawn eggs to work when sneaking, or force containers to open.

Finally, this also allows preventing any interaction at all without preventing block placement (by setting both to false). Since cancelling the event will typically prevent placement too (which might not be desired).

Side note: Blocks are now always synced when right-clicking on a block.
This differs from the previous behaviour, where the blocks were only synced when the action "failed".
However, since this change introduces a situation where the action may succeed but have different results than the client expects, it's best to just always sync blocks in this situation.

Fixes #3267

JavierLeon9966 committed:
Add support for basic entity picking via middle-click (#5397)

Support for more advanced stuff like NBT copying wasn't added in this PR, as the NBT used by PM is currently an inconsistent mess and doesn't play nice with vanilla. In the interests of avoiding this mess propagating, it's been left for another time.

Adds PlayerEntityPickEvent a la PlayerBlockPickEvent
and Entity->getPickedItem().

Commit:
Merge 'stable' into 'minor-next'

Automatic merge performed by: https://github.com/pmmp/RestrictedActions/actions/runs/11865975725

pmmp-admin-bot[bot] committed:
Merge 'stable' into 'minor-next'

Automatic merge performed by: https://github.com/pmmp/RestrictedActions/actions/runs/11866065217

ipad54 committed:
Implemented End Crystal (#4715)

Co-authored-by: Dylan T. <dktapps@pmmp.io>

ipad54 committed:
Added ability to pick end crystal item (#6509)

pmmp-admin-bot[bot] committed:
Merge 'stable' into 'minor-next'

Automatic merge performed by: https://github.com/pmmp/RestrictedActions/actions/runs/11874902000

pmmp-admin-bot[bot] committed:
Merge 'stable' into 'minor-next'

Automatic merge performed by: https://github.com/pmmp/RestrictedActions/actions/runs/11904511045

pmmp-admin-bot[bot] committed:
Merge 'stable' into 'minor-next'

Automatic merge performed by: https://github.com/pmmp/RestrictedActions/actions/runs/11924942897

dktapps committed:
ConsoleReader: Use proc_open()'s socket support to send commands back to the main server process (#5273)

Support for this was introduced in PHP 8.0, though not mentioned in any changelog: php/php-src#5777

This simplifies the subprocess handling considerably. However, there is a potential for problems if PHP generates any E_* errors, since these get written to STDOUT as well.
To avoid error messages being treated as a command, a hash is attached to each IPC message, seeded with an incrementing counter. This prevents error messages causing command replays or unintended commands.

Unfortunately, PHP doesn't support binding pipes other than stdin/stdout/stderr on Windows for the child process, so we have to use stdout for this. In the future, if it becomes possible, a dedicated pipe for the purpose should be introduced. We'd need something like php://fd/<number> to work on Windows.

dktapps committed:
Move event handler inheritance test to PHPUnit using mock objects
this is dodgy and we shouldn't rely on it long term.

relates to #6524

dktapps committed:
Move event test fixtures to a subdirectory

dktapps committed:
Add setUp and tearDown for event unit tests

pmmp-admin-bot[bot] committed:
Merge 'stable' into 'minor-next'

Automatic merge performed by: https://github.com/pmmp/RestrictedActions/actions/runs/11964819241

pmmp-admin-bot[bot] committed:
Merge 'stable' into 'minor-next'

Automatic merge performed by: https://github.com/pmmp/RestrictedActions/actions/runs/11989715327

pmmp-admin-bot[bot] committed:
Merge 'stable' into 'minor-next'

Automatic merge performed by: https://github.com/pmmp/RestrictedActions/actions/runs/11990103798

pmmp-admin-bot[bot] committed:
Merge 'stable' into 'minor-next'

Automatic merge performed by: https://github.com/pmmp/RestrictedActions/actions/runs/11990667732

ShockedPlot7560 committed:
Add generic types for TaskHandler (#6030)

zSALLAZAR committed:
Implement Ice Bomb (#5452)

Co-authored-by: Dylan K. Taylor <dktapps@pmmp.io>

dktapps committed:
Merge branch 'stable' into minor-next

dktapps committed:
Merge branch 'stable' into minor-next

dktapps committed:
PermissionManager: deprecate permission subscription system

this is no longer used by the core, and as far as I can tell no plugin uses it either.
it was used in the past for chat broadcast channels, but not anymore.

dktapps committed:
Deprecate InventoryAction->onAddToTransaction()
this never made any sense

pmmp-admin-bot[bot] committed:
Merge 'stable' into 'minor-next'

Automatic merge performed by: https://github.com/pmmp/RestrictedActions/actions/runs/12022106146

AkmalFairuz committed:
Introduce Utils::getRandomFloat() (#6532)

Drop-in replacement for lcg_value() for PHP 8.4

dktapps committed:
Merge branch 'stable' into minor-next

pmmp-admin-bot[bot] committed:
Merge 'stable' into 'minor-next'

Automatic merge performed by: https://github.com/pmmp/RestrictedActions/actions/runs/12060967026

pmmp-admin-bot[bot] committed:
Merge 'stable' into 'minor-next'

Automatic merge performed by: https://github.com/pmmp/RestrictedActions/actions/runs/12077579940

pmmp-admin-bot[bot] committed:
Merge 'stable' into 'minor-next'

Automatic merge performed by: https://github.com/pmmp/RestrictedActions/actions/runs/12092463227

dktapps committed:
Support for collecting timings from threads, and implement async task timings (#6333)

The following callbacks can now be registered in timings, to allow threads to be notified of these events:
- Turning on/off (`TimingsHandler::getToggleCallbacks()->add(...)`)
- Reset (`TimingsHandler::getReloadCallbacks()->add(...)`)
- Collect (`TimingsHandler::getCollectCallbacks()->add(...)`)

Collect callbacks must return `list<Promise>`. The promises must be `resolve()`d with `list<string>` of printed timings records, as returned by `TimingsHandler::printCurrentThreadRecords()`. It's recommended to use 1 promise per thread.

A timings report will be produced once all promises have been resolved.

This system is used internally to collect timings for async tasks (closes #6166).

For timings viewer developers:
Timings format version has been bumped to 3 to accommodate this change. Timings groups should now include a `ThreadId`  at the end of timings group names to ensure that their record IDs are segregated correctly, as they could otherwise conflict between threads. The main thread is not required to specify a thread ID. See pmmp/timings@13cefa6279ee8866dc584b9a7977a9c09cecd732 for implementation examples.

New PHPStan error is caused by phpstan/phpstan#10924

dktapps committed:
Deprecate some stuff

dktapps committed:
Added ClosureCommand (#6063)

this is intended to replace PluginCommand and CommandExecutor, both of which are overengineered and unfit for purpose.
Allowing a closure allows much greater flexibility.

We can't use this within the core yet, as plugins will expect PluginBase->getCommand() to return PluginCommand (with its associated setExecutor() and similar APIs).
However, I think this is useful enough to add by itself.

dktapps committed:
Allow eating in creative & peaceful
closes #5923
closes #6056

IvanCraft623 committed:
Implement Recovery compass (#5502)

Co-authored-by: Dylan K. Taylor <dktapps@pmmp.io>

IvanCraft623 committed:
Implement frost walker enchantment (#5497)

Co-authored-by: Dylan T. <dktapps@pmmp.io>

pmmp-admin-bot[bot] committed:
Merge 'stable' into 'minor-next'

Automatic merge performed by: https://github.com/pmmp/RestrictedActions/actions/runs/12111121061

pmmp-admin-bot[bot] committed:
Merge 'stable' into 'minor-next'

Automatic merge performed by: https://github.com/pmmp/RestrictedActions/actions/runs/12131296321

dktapps committed:
Switch back to official JsonMapper
the issues that led to the need for a fork have been addressed in the 5.0.0 release.

pmmp-admin-bot[bot] committed:
Merge 'stable' into 'minor-next'

Automatic merge performed by: https://github.com/pmmp/RestrictedActions/actions/runs/12151373979

dktapps committed:
Merge branch 'stable' of github.com:pmmp/PocketMine-MP into minor-next

dktapps committed:
AsyncTask: deprecate progress update related stuff

dktapps committed:
more deprecations in line with major-next

dktapps committed:
Prepare 5.23.0 release

ipad54 commented on changelogs/5.23.md:
```diff
@@ -0,0 +1,103 @@
+- `start.sh` will now emit warnings when the server process exits with an unusual exit code. This helps to detect unexpected segmentation faults and other kinds of native errors.
+
+## Gameplay
+- Added the following new items:
```
You haven’t included the ice bomb in this list

dktapps replied:
Good catch

dktapps committed:
Update 5.23.md

ipad54 commented on changelogs/5.23.md:
```diff
@@ -0,0 +1,104 @@
+  - `public static TimingsHandler::getReloadCallbacks() : ObjectSet<\Closure() : void>` - callbacks called when timings are reset
+  - `public static TimingsHandler::getToggleCallbacks() : ObjectSet<\Closure(bool $enable) : void>` - callbacks called when timings are enabled/disabled
+  - `public static TimingsHandler::requestPrintTimings() : Promise<list<string>>` - asynchronously collects timing results from all threads and assembles them into a single report
+- The following API methods have been deprecated:
```
I’ve also observed now that there is no mention about deprecation `Timings::getAsyncTaskErrorTimings()` and `Timings::GROUP_BREAKDOWN`

dktapps replied:
Those are internal, so I didn't think it was necessary to mention them. Really, the problem is that they've never been marked as such.

dktapps committed:
Mention internal timings deprecations
plugins shouldn't be using these, but since it's not marked as internal, we can't be sure.
